ec2/utils.py

```diff
@@ -76,120 +215,267 @@ def get_instance_state(instance, ctx):
-                                  'OS Returned: {1}'.format(
-                                      ctx.node.properties['private_key_path'],
-                                      OSError))
+        key = ec2_client.get_key_pair(ctx.node.properties['resource_id'])
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to get key pair: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+
+    return key.name
+
+
+def get_address_by_id(address_id, ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+    try:
+        address = ec2_client.get_all_addresses(address_id)
+    except boto.exception.EC2ResponseError as e:
+        report_all_addresses(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except (boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    return address.public_ip
+
+
+def get_address_object_by_id(address_id, ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+    try:
+        address = ec2_client.get_all_addresses(address_id)
+    except boto.exception.EC2ResponseError as e:
+        report_all_addresses(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except (boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to get address by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    return address
+
+
+def report_all_instances(ctx):
+    ec2_client = connection.EC2ConnectionClient().client()
+
+    try:
+        reservations = ec2_client.get_all_reservations()
+    except (boto.exception.EC2ResponseError,
+            boto.exception.BotoServerError) as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to report all instances: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+
+    instances = [instance for res in reservations
+                 for instance in res.instances]
+
+    message = []
+    for instance in instances:
+        message.append('{}\n'.format(instance))
+
+    ctx.logger.info('Available instances: {0}'.format(message))
+
+
+def report_all_security_groups(ctx):
```

Comment:
Changed to get_all_security_groups and if it fails it runs 'log_available_resources' and then raises non-recoverable